File: com/jetbrains/teamsys/dnq/database/TransientEntityStoreImpl.java
package com.jetbrains.teamsys.dnq.database;

import jetbrains.exodus.core.dataStructures.StablePriorityQueue;
import jetbrains.exodus.core.dataStructures.hash.HashSet;
import jetbrains.exodus.core.execution.locks.Latch;
import jetbrains.exodus.database.TransientEntity;
import jetbrains.exodus.database.TransientEntityStore;
import jetbrains.exodus.database.TransientStoreSession;
import jetbrains.exodus.database.TransientStoreSessionListener;
import jetbrains.exodus.entitystore.*;
import jetbrains.exodus.entitystore.metadata.ModelMetaData;
import jetbrains.exodus.env.EnvironmentConfig;
import jetbrains.exodus.query.QueryEngine;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * @author Vadim.Gurov
 */
public class TransientEntityStoreImpl implements TransientEntityStore {

    private static final Log log = LogFactory.getLog(TransientEntityStoreImpl.class);

    private EntityStore persistentStore;
    private QueryEngine queryEngine;
    private ModelMetaData modelMetaData;
    private IEventsMultiplexer eventsMultiplexer;
    private final Set<TransientStoreSession> sessions = new HashSet<TransientStoreSession>(200);
    private final ThreadLocal<TransientStoreSession> currentSession = new ThreadLocal<TransientStoreSession>();
    private final StablePriorityQueue<Integer, TransientStoreSessionListener> listeners = new StablePriorityQueue<Integer, TransientStoreSessionListener>();

    private boolean open = true;
    private final Latch enumContainersLock = Latch.create();
    private final Set<EnumContainer> initedContainers = new HashSet<EnumContainer>(10);
    private final Map<String, Entity> enumCache = new ConcurrentHashMap<String, Entity>();
    private final Map<String, BasePersistentClassImpl> persistentClassInstanceCache = new ConcurrentHashMap<String, BasePersistentClassImpl>();
    private final Map<Class, BasePersistentClassImpl> persistentClassInstances = new ConcurrentHashMap<Class, BasePersistentClassImpl>();

    public TransientEntityStoreImpl() {
        if (log.isTraceEnabled()) {
            log.trace("TransientEntityStoreImpl constructor called.");
        }
    }

    public EntityStore getPersistentStore() {
        return persistentStore;
    }

    public QueryEngine getQueryEngine() {
        return queryEngine;
    }

    public IEventsMultiplexer getEventsMultiplexer() {
        return eventsMultiplexer;
    }

    public void setEventsMultiplexer(IEventsMultiplexer eventsMultiplexer) {
        this.eventsMultiplexer = eventsMultiplexer;
    }

    /**
     * Service locator {@link jetbrains.springframework.configuration.runtime.ServiceLocator} is responsible to set persistent entity store
     *
     * @param persistentStore persistent entity store.
     */
    @SuppressWarnings({"UnusedDeclaration"})
    public void setPersistentStore(EntityStore persistentStore) {
        this.persistentStore = persistentStore;
    }

    /**
     * Service locator {@link jetbrains.springframework.configuration.runtime.ServiceLocator} is responsible to set query engine
     *
     * @param persistentStore persistent entity store.
     */
    @SuppressWarnings({"UnusedDeclaration"})
    public void setQueryEngine(QueryEngine queryEngine) {
        this.queryEngine = queryEngine;
    }

    @NotNull
    public String getName() {
        return "transient store";
    }

    @NotNull
    public String getLocation() {
        throw new UnsupportedOperationException("Not supported by transient store.");
    }

    @NotNull
    @Override
    public StoreTransaction beginTransaction() {
        throw new UnsupportedOperationException();
    }

    @NotNull
    @Override
    public StoreTransaction beginReadonlyTransaction() {
        return registerStoreSession(new TransientSessionImpl(this, true));
    }

    @Nullable
    @Override
    public StoreTransaction getCurrentTransaction() {
        throw new UnsupportedOperationException();
    }

    public TransientStoreSession beginSession() {
        assertOpen();

        if (log.isDebugEnabled()) {
            log.debug("Begin new session");
        }

        TransientStoreSession currentSession = this.currentSession.get();
        if (currentSession != null) {
            log.debug("Return session already associated with the current thread " + currentSession);
            return currentSession;
        }

        return registerStoreSession(new TransientSessionImpl(this, false));
    }

    public void resumeSession(TransientStoreSession session) {
        if (session != null) {
            assertOpen();

            TransientStoreSession current = currentSession.get();
            if (current != null) {
                if (current != session) {
                    throw new IllegalStateException("Another open transient session already associated with current thread.");
                }
            }

            currentSession.set(session);
        }
    }

    public void setModelMetaData(final ModelMetaData modelMetaData) {
        this.modelMetaData = modelMetaData;
    }

    @Nullable
    public ModelMetaData getModelMetaData() {
        return modelMetaData;
    }

    /**
     * It's guaranteed that current thread session is Open, if exists
     *
     * @return current thread session
     */
    @Nullable
    public TransientStoreSession getThreadSession() {
        return currentSession.get();
    }

    public void close() {
        eventsMultiplexer.onClose(this);

        log.info("Close transient store.");
        open = false;

        int sessionsSize = sessions.size();
        if (sessionsSize > 0) {
            log.warn("There're " + sessionsSize + " open transient sessions. Print.");
            if (log.isDebugEnabled()) {
                for (TransientStoreSession session : sessions) {
                    TransientSessionImpl impl = session instanceof TransientSessionImpl ? (TransientSessionImpl) session : null;
                    if (impl != null) {
                        log.warn("Not closed session stack trace: ", impl.getStack());
                    }
                }
            }
        }
    }

    public boolean entityTypeExists(@NotNull final String entityTypeName) {
        try {
            return ((PersistentEntityStore) persistentStore).getEntityTypeId(entityTypeName, false) >= 0;
        } catch (Exception e) {
            // ignore
        }
        return false;
    }

    public void renameEntityTypeRefactoring(@NotNull final String oldEntityTypeName, @NotNull final String newEntityTypeName) {
        final TransientSessionImpl s = (TransientSessionImpl) getThreadSession();

        if (s == null) {
            throw new IllegalStateException("No current thread session.");
        }

        s.addChangeAndRun(new TransientSessionImpl.MyRunnable() {
            @Override
            public boolean run() {
                ((PersistentEntityStore) s.getPersistentTransaction().getStore()).renameEntityType(oldEntityTypeName, newEntityTypeName);
                return true;
            }
        });
    }

    public void deleteEntityTypeRefactoring(@NotNull final String entityTypeName) {
        final TransientSessionImpl s = (TransientSessionImpl) getThreadSession();

        if (s == null) {
            throw new IllegalStateException("No current thread session.");
        }

        s.addChangeAndRun(new TransientSessionImpl.MyRunnable() {
            @Override
            public boolean run() {
                ((PersistentEntityStoreImpl) s.getPersistentTransaction().getStore()).deleteEntityType(entityTypeName);
                return true;
            }
        });
    }

    public void deleteEntityRefactoring(@NotNull Entity entity) {
        final TransientSessionImpl s = (TransientSessionImpl) getThreadSession();

        if (s == null) {
            throw new IllegalStateException("No current thread session.");
        }

        final Entity persistentEntity = (entity instanceof TransientEntity) ? ((TransientEntity) entity).getPersistentEntity() : entity;
        if (entity instanceof TransientEntity) {
            s.deleteEntity((TransientEntity) entity);
        } else {
            s.addChangeAndRun(new TransientSessionImpl.MyRunnable() {
                public boolean run() {
                    persistentEntity.delete();
                    return true;
                }
            });
        }
    }

    public void deleteLinksRefactoring(@NotNull final Entity entity, @NotNull final String linkName) {
        final TransientSessionImpl s = (TransientSessionImpl) getThreadSession();

        if (s == null) {
            throw new IllegalStateException("No current thread session.");
        }

        final Entity persistentEntity = (entity instanceof TransientEntity) ? ((TransientEntity) entity).getPersistentEntity() : entity;
        s.addChangeAndRun(new TransientSessionImpl.MyRunnable() {
            public boolean run() {
                persistentEntity.deleteLinks(linkName);
                return true;
            }
        });
    }

    public void deleteLinkRefactoring(@NotNull final Entity entity, @NotNull final String linkName, @NotNull final Entity link) {
        final TransientSessionImpl s = (TransientSessionImpl) getThreadSession();

        if (s == null) {
            throw new IllegalStateException("No current thread session.");
        }

        final Entity persistentEntity = (entity instanceof TransientEntity) ? ((TransientEntity) entity).getPersistentEntity() : entity;
        final Entity persistentLink = (link instanceof TransientEntity) ? ((TransientEntity) link).getPersistentEntity() : link;

        s.addChangeAndRun(new TransientSessionImpl.MyRunnable() {
            public boolean run() {
                persistentEntity.deleteLink(linkName, persistentLink);
                return true;
            }
        });
    }

    private TransientStoreSession registerStoreSession(TransientStoreSession s) {
        synchronized (sessions) {
            if (!sessions.add(s)) {
                throw new IllegalArgumentException("Session is already registered.");
            }
        }

        currentSession.set(s);

        return s;
    }

    void unregisterStoreSession(TransientStoreSession s) {
        synchronized (sessions) {
            if (!sessions.remove(s)) {
                throw new IllegalArgumentException("Transient session wasn't previously registered.");
            }
        }

        currentSession.remove();
    }

    @Nullable
    public TransientStoreSession suspendThreadSession() {
        assertOpen();

        final TransientStoreSession current = getThreadSession();
        if (current != null) {
            currentSession.remove();
        }

        return current;
    }

    public void addListener(@NotNull TransientStoreSessionListener listener) {
        listeners.push(Integer.valueOf(0), listener);
    }

    @Override
    public void addListener(TransientStoreSessionListener listener, final int priority) {
        listeners.push(Integer.valueOf(priority), listener);
    }

    public void removeListener(@NotNull TransientStoreSessionListener listener) {
        listeners.remove(listener);
    }

    void forAllListeners(@NotNull ListenerVisitor v) {
        for (final TransientStoreSessionListener listener : listeners) {
            v.visit(listener);
        }
    }

    public int sessionsCount() {
        synchronized (sessions) {
            return sessions.size();
        }
    }

    public void dumpSessions(StringBuilder sb) {
        synchronized (sessions) {
            for (TransientStoreSession s : sessions) {
                sb.append("\n").append(s.toString());
            }
        }
    }

    public boolean isEnumContainerInited(EnumContainer container) {
        return initedContainers.contains(container);
    }

    public void enumContainerInited(EnumContainer container) {
        initedContainers.add(container);
    }

    public void enumContainerLock() throws InterruptedException {
        enumContainersLock.acquire();
    }

    public void enumContainerUnLock() {
        enumContainersLock.release();
    }

    public Entity getCachedEnumValue(@NotNull final String className, @NotNull final String propName) {
        return enumCache.get(getEnumKey(className, propName));
    }

    public void setCachedEnumValue(@NotNull final String className,
                                   @NotNull final String propName, @NotNull final Entity entity) {
        enumCache.put(getEnumKey(className, propName), entity);
    }

    public BasePersistentClassImpl getCachedPersistentClassInstance(@NotNull final String entityType) {
        return persistentClassInstanceCache.get(entityType);
    }

    public BasePersistentClassImpl getCachedPersistentClassInstance(@NotNull final Class<? extends BasePersistentClassImpl> entityType) {
        return persistentClassInstances.get(entityType);
    }

    public void setCachedPersistentClassInstance(@NotNull final String entityType, @NotNull final BasePersistentClassImpl instance) {
        persistentClassInstanceCache.put(entityType, instance);
        Class<? extends BasePersistentClassImpl> clazz = instance.getClass();
        if (persistentClassInstances.get(clazz) != null) {
            if (log.isWarnEnabled()) {
                log.warn("Persistent class instance already registered for: " + clazz.getSimpleName());
            }
        }
        persistentClassInstances.put(clazz, instance);
    }

    private void assertOpen() {
        if (!open) throw new IllegalStateException("Transient store is closed.");
    }

    public static String getEnumKey(@NotNull final String className, @NotNull final String propName) {
        final StringBuilder builder = new StringBuilder(24);
        builder.append(propName);
        builder.append('@');
        builder.append(className);
        return builder.toString();
    }

    interface ListenerVisitor {
        void visit(TransientStoreSessionListener listener);
    }

}

File: dnq-transient-store/src/main/kotlin/com/jetbrains/teamsys/dnq/database/TransientEntityStoreImpl.kt
/**
 * Copyright 2006 - 2024 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jetbrains.teamsys.dnq.database

import jetbrains.exodus.core.dataStructures.StablePriorityQueue
import jetbrains.exodus.database.*
import jetbrains.exodus.entitystore.Entity
import jetbrains.exodus.entitystore.PersistentEntityStore
import jetbrains.exodus.entitystore.QueryCancellingPolicy
import jetbrains.exodus.entitystore.StoreTransaction
import jetbrains.exodus.entitystore.orientdb.OPersistentEntityStore
import jetbrains.exodus.entitystore.orientdb.OVertexEntity
import jetbrains.exodus.entitystore.util.unsupported
import jetbrains.exodus.env.EnvironmentConfig
import jetbrains.exodus.query.QueryEngine
import jetbrains.exodus.query.metadata.ModelMetaData
import mu.KLogging
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * @author Vadim.Gurov
 */
open class TransientEntityStoreImpl : TransientEntityStore {

    companion object : KLogging()

    private lateinit var _persistentStore: OPersistentEntityStore

    var entityLifecycle: EntityLifecycle? = null

    /**
     * Must be injected.
     */
    override var persistentStore: OPersistentEntityStore
        get() = _persistentStore
        set(persistentStore) {
            val ec = persistentStore.environment.environmentConfig
            if (ec.envTxnDowngradeAfterFlush == EnvironmentConfig.DEFAULT.envTxnDowngradeAfterFlush) {
                ec.envTxnDowngradeAfterFlush = false
            }
            ec.envTxnReplayMaxCount = Integer.MAX_VALUE
            ec.envTxnReplayTimeout = Long.MAX_VALUE
            ec.gcUseExclusiveTransaction = true
            _persistentStore = persistentStore
        }

    /**
     * Must be injected.
     */
    override lateinit var queryEngine: QueryEngine
    override var modelMetaData: ModelMetaData? = null
    override var changesMultiplexer: ITransientChangesMultiplexer? = null
        set(value) {
            field = value
            if (value is TransientStoreSessionListener) {
                addListener(value, Int.MAX_VALUE)
            }
        }

    override var invocationTransport: ListenerInvocationTransport? = null

    private val sessions = Collections.newSetFromMap(ConcurrentHashMap<TransientStoreSession, Boolean>(200))
    private val currentSession = ThreadLocal<TransientStoreSession>()
    private val listeners = StablePriorityQueue<Int, TransientStoreSessionListener>()

    @field:Volatile
    override var isOpen = true
        protected set
    private var closed = false
    private val enumCache = ConcurrentHashMap<String, Entity>()

    /**
     * It's guaranteed that the current thread session is Open, if exists
     */
    override val threadSession: TransientStoreSession?
        get() = currentSession.get()

    private val transientSessionOrThrow: TransientSessionImpl
        get() = threadSessionOrThrow as TransientSessionImpl

    override fun getName() = persistentStore.name

    override fun getLocation(): String = persistentStore.location

    override fun <T> transactional(
            readonly: Boolean,
            queryCancellingPolicy: QueryCancellingPolicy?,
            isNew: Boolean,
            block: (TransientStoreSession) -> T
    ): T = TransientEntityStoreExt.transactional(
        this, queryCancellingPolicy, isNew, block)

    override fun beginTransaction(): StoreTransaction {
        throw UnsupportedOperationException()
    }

    override fun beginExclusiveTransaction(): StoreTransaction {
        throw UnsupportedOperationException()
    }

    override fun beginReadonlyTransaction(): TransientStoreSession {
        return registerStoreSession(TransientSessionImpl(this, readonly = true))
    }

    override fun getCurrentTransaction(): StoreTransaction? {
        throw UnsupportedOperationException()
    }

    override fun beginSession(): TransientStoreSession {
        assertOpen()

        logger.debug { "Begin new session" }

        val currentSession = this.currentSession.get()
        if (currentSession != null) {
            logger.debug { "Return session already associated with the current thread $currentSession" }
            return currentSession
        }

        return registerStoreSession(TransientSessionImpl(this, readonly = false))
    }

    override fun resumeSession(session: TransientStoreSession?) {
        if (session != null) {
            assertOpen()

            val current = currentSession.get()
            if (current != null && current != session) {
                throw IllegalStateException("Another open transient session is already associated with current thread")
            }
            currentSession.set(session)
        }
    }

    override fun close() {
        isOpen = false

        changesMultiplexer?.onClose(this)

        logger.debug { "Close transient store." }
        closed = true

        val sessionsSize = sessions.size
        if (sessionsSize > 0) {
            logger.warn { "There're $sessionsSize open transient sessions. Print." }
            if (logger.isDebugEnabled) {
                sessions.asSequence()
                        .filterIsInstance<TransientSessionImpl>()
                        .mapNotNull { session -> session.stack }
                        .forEach { sessionStackTrace ->
                            logger.warn(sessionStackTrace) { "Not closed session stack trace: " }
                        }
            }
        }
        _persistentStore.close()
        _persistentStore.environment.close()
    }


    override fun entityTypeExists(entityTypeName: String): Boolean {
        return try {
            _persistentStore.getEntityTypeId(entityTypeName) >= 0
        } catch (e: Exception) {
            false
        }
    }

    override fun renameEntityTypeRefactoring(oldEntityTypeName: String, newEntityTypeName: String) {
        val transientSession = transientSessionOrThrow
        transientSession.entitiesUpdater.addChangeAndRun {
            (transientSession.transactionInternal.store as PersistentEntityStore).renameEntityType(oldEntityTypeName, newEntityTypeName)
            true
        }
    }

    override fun deleteEntityTypeRefactoring(entityTypeName: String) {
        unsupported { "This should be done with low level in orientDB" }
    }

    override fun deleteEntityRefactoring(entity: Entity) {
        val transientSession = transientSessionOrThrow

        if (entity is TransientEntity) {
            transientSession.entitiesUpdater.deleteEntity(entity)
        } else {
            val persistentEntity = entity.unwrapEntity()
            transientSession.entitiesUpdater.addChangeAndRun {
                persistentEntity.delete()
                true
            }
        }
    }

    override fun deleteLinksRefactoring(entity: Entity, linkName: String) {
        val transientSession = transientSessionOrThrow

        val persistentEntity = entity.unwrapEntity()
        transientSession.entitiesUpdater.addChangeAndRun {
            persistentEntity.deleteLinks(linkName)
            true
        }
    }

    override fun deleteLinkRefactoring(entity: Entity, linkName: String, link: Entity) {
        val transientSession = transientSessionOrThrow

        val persistentEntity = entity.unwrapEntity()
        val persistentLink = link.unwrapEntity()

        transientSession.entitiesUpdater.addChangeAndRun {
            persistentEntity.deleteLink(linkName, persistentLink)
            true
        }
    }

    fun registerStoreSession(storeSession: TransientStoreSession): TransientStoreSession {
        if (!sessions.add(storeSession)) {
            throw IllegalArgumentException("Session is already registered.")
        }
        currentSession.set(storeSession)
        return storeSession
    }

    fun unregisterStoreSession(storeSession: TransientStoreSession) {
        if (!sessions.remove(storeSession)) {
            throw IllegalArgumentException("Transient session wasn't previously registered.")
        }
        currentSession.remove()
    }

    override fun suspendThreadSession(): TransientStoreSession? {
        assertOpen()

        val current = threadSession
        if (current != null) {
            currentSession.remove()
        }

        return current
    }

    override fun addListener(listener: TransientStoreSessionListener) {
        addListener(listener, 0)
    }

    override fun addListener(listener: TransientStoreSessionListener, priority: Int) {
        withListeners {
            if (any { it == listener }) {
                throw IllegalStateException("$listener is already registered as a listener")
            }
            push(priority, listener)
        }
    }

    override fun removeListener(listener: TransientStoreSessionListener) {
        withListeners {
            remove(listener)
        }
    }

    internal fun forAllListeners(action: (TransientStoreSessionListener) -> Unit) {
        withListeners {
            toList()
        }.forEach(action)
    }

    fun sessionsCount(): Int {
        return sessions.size
    }


    fun dumpSessions(sb: StringBuilder) {
        sessions.joinTo(sb, "\n")
    }

    override fun getCachedEnumValue(className: String, propName: String): Entity? {
        return enumCache[getEnumKey(className, propName)]
    }

    fun setCachedEnumValue(className: String, propName: String, entity: Entity) {
        enumCache[getEnumKey(className, propName)] = entity
    }

    private fun Entity.unwrapEntity(): OVertexEntity {
        return when (this) {
            is TransientEntity -> this.entity as OVertexEntity
            is OVertexEntity -> this
            else -> throw IllegalArgumentException("Cannot unwrap entity")
        }
    }

    private fun <T> withListeners(action: StablePriorityQueue<Int, TransientStoreSessionListener>.() -> T): T {
        listeners.lock()
        try {
            return action(listeners)
        } finally {
            listeners.unlock()
        }
    }

    private fun getEnumKey(className: String, propName: String) = "$propName@$className"

    private fun assertOpen() {
        // this flag isn't even volatile, but this is legacy behavior
        if (closed) throw IllegalStateException("Transient store is closed.")
    }
}

